### DISCUSSION TITLE
Go datasource cannot work with responses from /list containing spaces at the end anymore

### DISCUSSION THREAD

**Discussion by ArtAndreev** [renovatebot/renovate#27178] [Request Help]
### How are you running Renovate?

Self-hosted Renovate

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.179.0

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

It used to work, and then stopped

### Describe the problem

Renovate worked okay with go datasource, but suddenly stopped updating regular dependencies about two weeks ago, the only dependencies to update have digest versions.

I've investigated the problem.

First, I opened logs and saw requests to our proxy in wrong format, they contain spaces (for example, `"https://private-go-proxy/github.com/google/uuid/@v/v1.6.0 .info (GET,404)": 1`, see logs below). This proxy is a quite popular vendor solution for go proxy.

Second, I saw `200 OK` responses to `/list` requests. I requested this URL and got the next response:

<img width="129" alt="image" src="https://github.com/renovatebot/renovate/assets/26388769/7c8859b2-3ad4-4e0f-9862-b6eca4c03e0d">

As you see, it contains spaces at the end of each line. This is not 100%-valid, but before we had no problem with it! Also `go get` doesn't fail with this proxy, it sanitizes lines (see [permalink to `go get` implementation](https://github.com/golang/go/blob/2057ad02bd8387378a2d1fd637e955e126f698bf/src/cmd/go/internal/modfetch/proxy.go#L304)).

At last, I thought the problem was in renovate itself, I cloned the repository, found a file with a request to `/list`: [see](https://github.com/renovatebot/renovate/blob/f69d7e83b33727cb66742facafb24e1c3ce1e432/lib/modules/datasource/go/releases-goproxy.ts#L212). I ran `git blame` and found the commit that introduced the issue and its PR: [see lines](https://github.com/renovatebot/renovate/pull/26746/files#diff-300928c3d31ec1aa23ceceeb509e51f34eaea798ca3d2a866899a13e1f3c2809L214-R235). I attach a part of `diff` output with problematic lines below:

<details><summary>diff for commit</summary>

```diff
diff --git a/lib/modules/datasource/go/releases-goproxy.ts b/lib/modules/datasource/go/releases-goproxy.ts
index 82fa68f43..db6db3d94 100644
--- a/lib/modules/datasource/go/releases-goproxy.ts
+++ b/lib/modules/datasource/go/releases-goproxy.ts
@@ -211,10 +215,24 @@ export class GoProxyDatasource extends Datasource {
       .split(newlineRegex)
       .filter(is.nonEmptyStringAndNotWhitespace)
       .map((str) => {
-        const [version, releaseTimestamp] = str.split(regEx(/\s+/));
-        return DateTime.fromISO(releaseTimestamp).isValid
-          ? { version, releaseTimestamp }
-          : { version };
+        let version: string | undefined;
+        let newDigest: string | undefined;
+        let releaseTimestamp: string | undefined;
+        [version, releaseTimestamp] = str.split(regEx(/\s+/));
+
+        if (!DateTime.fromISO(releaseTimestamp).isValid) {
+          version = str;
+          releaseTimestamp = undefined;
+        }
+        const digestMatch = pversionRegexp.exec(version);
+
+        if (!releaseTimestamp && digestMatch?.groups?.releaseTimestamp) {
+          releaseTimestamp = digestMatch.groups.releaseTimestamp;
+        }
+        if (digestMatch?.groups?.digest) {
+          newDigest = digestMatch.groups.digest;
+        }
+        return { version, newDigest, releaseTimestamp };
       });
   }
```

</details>

The version is without timestamp. Old implementation returned `version` after splitting, that had removed spaces, `: { version };`, but the new one overrides version to raw value, `version = str;`, this removes space trimming and causes troubles described above.

So the fix is in correct sanitizing of `version` when timestamp is missing in response.

Can't create a reproduction repo since I can't deploy all our private infrastructure.

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: http statistics (repository=private-repository)
       "urls": {
         "https://private-go-proxy/github.com/google/uuid/@latest (GET,200)": 1,
         "https://private-go-proxy/github.com/google/uuid/@v/list (GET,200)": 1,
         "https://private-go-proxy/github.com/google/uuid/@v/v1.0.0 .info (GET,404)": 1,
         "https://private-go-proxy/github.com/google/uuid/@v/v1.1.0 .info (GET,404)": 1,
         "https://private-go-proxy/github.com/google/uuid/@v/v1.1.1 .info (GET,404)": 1,
         "https://private-go-proxy/github.com/google/uuid/@v/v1.1.2 .info (GET,404)": 1,
         "https://private-go-proxy/github.com/google/uuid/@v/v1.1.3 .info (GET,404)": 1,
         "https://private-go-proxy/github.com/google/uuid/@v/v1.1.4 .info (GET,404)": 1,
         "https://private-go-proxy/github.com/google/uuid/@v/v1.1.5 .info (GET,404)": 1,
         "https://private-go-proxy/github.com/google/uuid/@v/v1.2.0 .info (GET,404)": 1,
         "https://private-go-proxy/github.com/google/uuid/@v/v1.3.0 .info (GET,404)": 1,
         "https://private-go-proxy/github.com/google/uuid/@v/v1.3.1 .info (GET,404)": 1,
         "https://private-go-proxy/github.com/google/uuid/@v/v1.4.0 .info (GET,404)": 1,
         "https://private-go-proxy/github.com/google/uuid/@v/v1.5.0 .info (GET,404)": 1,
         "https://private-go-proxy/github.com/google/uuid/@v/v1.6.0 .info (GET,404)": 1,
         "https://private-go-proxy/github.com/google/uuid/v2/@v/list (GET,404)": 1,
```

</details>

### Have you created a minimal reproduction repository?

I have explained in the description why a minimal reproduction is impossible

**Comment by viceice** [comment]
@zharinov looks like a bug from refactor. can you fix it?

**Reply by zharinov** [reply]
Ok

**Comment by ArtAndreev** [comment]
Just checked out, it works, thanks!
